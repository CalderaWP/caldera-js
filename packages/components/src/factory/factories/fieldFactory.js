import {
	RemotePost,
	TextAreaField,
	ToggleField,
	RadioField,
	FieldWrapper,
	FieldSet,
	SelectField,
	InputField,
	RichText,
	MagicRichText,
	isValidHtml5type,
	AutoCompleteField,
	FormFieldsAutoComplete
} from '@calderawp/components';
import React from 'react';

/**
 * Create a field from an object describing it.
 *
 * Note: field.render is a render prop. If it is passed, field is passed to it.
 *
 * @param field
 * @param onChange
 * @param onBlur
 * @return {*}
 */
export const fieldFactory = (field, onChange, onBlur) => {
	const {
		fieldType,
		label,
		attributes,
		options,
		fieldId,
		messages,
		render
	} = field;

	if (render) {
		return React.createElement(render, field);
	}

	switch (fieldType) {
		case 'checkboxes':
			return (
				<FieldSet
					fieldType={fieldType}
					legend={label}
					attributes={attributes}
				>
					{options.map(option => {
						const {
							value,
							label,
							description,
							attributes
						} = option;
						const optionId = option.hasOwnProperty('id')
							? option.id
							: `opt-${fieldId}-${option.value}`;
						return (
							<InputField
								key={optionId}
								id={optionId}
								fieldId={optionId}
								value={value}
								label={label}
								description={description}
								html5type={'checkbox'}
								attributes={attributes}
								onChange={onChange}
							/>
						);
					})}
				</FieldSet>
			);
		case 'magic-richtext' :
			return <MagicRichText {...field} onChange={onChange}/>;
		case 'autocomplete' :
			return <AutoCompleteField {...field} onChange={onChange}/>;
		case'richtext' :
			return <RichText {...field} onChange={onChange}/>;
			case'fields-autocomplete' :
			return <FormFieldsAutoComplete {...field} onChange={onChange}/>;
		case 'textarea':
			return <TextAreaField {...field} onChange={onChange}/>;
		case 'toggle':
			return <ToggleField {...field} onChange={onChange}/>;
		case 'radio':
			return <RadioField {...field} onChange={onChange}/>;
		case 'select':
		case 'dropdown':
			return <SelectField {...field} onChange={onChange}/>;
		case 'text':
		case 'email':
		case 'number':
		case 'input':
		default:
			if (isValidHtml5type(fieldType)) {
				if (field.html5type !== fieldType) {
					field = {...field, html5type: fieldType};
				}
			} else {
				field.html5type = 'text';
			}
			return <InputField {...field} onChange={onChange}/>;
	}
};
