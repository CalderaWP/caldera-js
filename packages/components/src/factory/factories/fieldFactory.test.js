import {fieldFactory} from './fieldFactory';
import React from 'react';
import renderer from 'react-test-renderer';
import {mount} from 'enzyme';
import {
	checkboxFieldset,
	selectField,
	checkboxField,
	numberField,
	textField,
	emailField,
	radioField,
	toggleField,
	textAreaField,
	autoCompleteField
} from '../fields.fixtures';

describe('fieldFactory', () => {
	let onChange;
	let onBlur;
	beforeEach(() => {
		onChange = jest.fn();
		onBlur = jest.fn();
	});
	it('Creates a text field', () => {
		const component = renderer.create(
			fieldFactory(textField, onChange, onBlur)
		);
		expect(component.toJSON()).toMatchSnapshot();
	});

	it('Creates a text field when field type is bogus', () => {
		const component = mount(
			fieldFactory(
				{...textField, fieldType: 'travels-in-space'},
				onChange,
				onBlur
			)
		);
		expect(component.find('input').prop('type')).toBe('text');
	});
	it('Creates a text field when field type is input', () => {
		const component = mount(
			fieldFactory({...textField, fieldType: 'input'}, onChange, onBlur)
		);
		expect(component.find('input').prop('type')).toBe('text');
	});

	it('Creates a number field', () => {
		const component = renderer.create(
			fieldFactory(numberField, onChange, onBlur)
		);
		expect(component.toJSON()).toMatchSnapshot();
	});

	it('Creates an email field', () => {
		const component = renderer.create(
			fieldFactory(emailField, onChange, onBlur)
		);
		expect(component.toJSON()).toMatchSnapshot();
	});

	it('Creates an checkbox field field', () => {
		const component = renderer.create(
			fieldFactory(checkboxField, onChange, onBlur)
		);
		expect(component.toJSON()).toMatchSnapshot();
	});

	it('Creates an checkbox field set', () => {
		const component = renderer.create(
			fieldFactory(checkboxFieldset, onChange, onBlur)
		);
		expect(component.toJSON()).toMatchSnapshot();
	});

	it('Creates an select field', () => {
		const component = renderer.create(
			fieldFactory(selectField, onChange, onBlur)
		);
		expect(component.toJSON()).toMatchSnapshot();
	});

	it('Creates a radio field', () => {
		const component = renderer.create(
			fieldFactory(radioField, onChange, onBlur)
		);
		expect(component.toJSON()).toMatchSnapshot();
	});

	it('Creates a checkbox fieldset', () => {
		const component = renderer.create(
			fieldFactory(checkboxFieldset, onChange, onBlur)
		);
		expect(component.toJSON()).toMatchSnapshot();
	});
	it('Creates an autocomplete field', () => {
		const component = renderer.create(
			fieldFactory(autoCompleteField, onChange, onBlur)
		);
		expect(component.toJSON()).toMatchSnapshot();
	});

	it('Creates an fields-autocomplete field', () => {
		const field = {
			...textField,
			fieldType: 'fields-autocomplete',
			form: {
				fields:[textField,textAreaField]
			}
		}

		const component = renderer.create(
			fieldFactory(field, onChange, onBlur)
		);
		expect(component.toJSON()).toMatchSnapshot();
	});

	it('Adds option id', () => {
		const component = mount(
			fieldFactory(
				{
					fieldType: 'checkboxes',
					label: 'Checkbox Field Set Label',
					fieldId: 'checkboxFieldSetLabel',
					description: 'Checkbox field set description',
					options: [
						{
							value: 1,
							label: 'One',
							id: 'opt-1'
						}
					]
				},
				onChange,
				onBlur
			)
		);
		expect(
			component
				.find('input')
				.first()
				.props().id
		).toBe('opt-1');
	});

	it('Adds option id based on fieldId if option id not provided', () => {
		const component = mount(
			fieldFactory(
				{
					fieldType: 'checkboxes',
					label: 'Checkbox Field Set Label',
					fieldId: 'test',
					description: 'Checkbox field set description',
					options: [
						{
							value: 7,
							label: 'One'
						}
					]
				},
				onChange,
				onBlur
			)
		);
		expect(
			component
				.find('input')
				.first()
				.props().id
		).toBe('opt-test-7');
	});

	it('Creates an select field identified as dropdown', () => {
		const component = renderer.create(
			fieldFactory({
				...selectField,
				fieldType: 'dropdown'
			}),
			onChange,
			onBlur
		);
		expect(component.toJSON()).toMatchSnapshot();
	});

	test('Render props pattern', () => {
		const _Field = props => <input type={'number'} key={88}/>;
		const _Rp = ({render}) => ({render});

		const component = mount(<_Rp render={_Field}/>);
		expect(component.find('input').prop('type')).toEqual('number');
	});

	test('Field can supply a component', () => {
		const _Field = props => <input type={'number'} key={88}/>;

		const component = mount(
			fieldFactory({
				render: _Field
			}),
			onChange,
			onBlur
		);
		expect(component.find('input').prop('type')).toEqual('number');
	});

	test('Field can supply a component and will be provided with field config as props', () => {
		const _Field = ({fieldId}) => (
			<input id={fieldId} type={'number'} key={88}/>
		);

		const component = mount(
			fieldFactory({
				...selectField,
				render: _Field
			}),
			onChange,
			onBlur
		);
		expect(component.find('#selectFieldId').length).toEqual(1);
	});

	it('should create textarea', () => {
		const component = renderer.create(
			fieldFactory(textAreaField, onChange, onBlur)
		);
		expect(component.toJSON()).toMatchSnapshot();
	});

	it('should create toggle', () => {
		const component = renderer.create(
			fieldFactory(toggleField, onChange, onBlur)
		);
		expect(component.toJSON()).toMatchSnapshot();
	});
});

describe('change handlers', () => {
	let onChange;
	let onBlur;
	beforeEach(() => {
		onChange = jest.fn();
		onBlur = jest.fn();
	});

	it('Changes calls change handler', () => {
		const component = mount(fieldFactory(textField, onChange, onBlur));
		component.find('input').simulate('change');
		expect(onChange.mock.calls.length).toBe(1);
	});
});
