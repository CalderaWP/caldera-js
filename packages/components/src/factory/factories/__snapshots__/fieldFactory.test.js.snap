// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fieldFactory Creates a checkbox fieldset 1`] = `
<fieldset
  className="caldera-fieldset caldera-fieldset-checkboxes"
>
  <legend>
    Checkbox Field Set Label
  </legend>
  <div
    className="components-base-control"
  >
    <div
      className="components-base-control__field"
    >
      <input
        aria-describedby={undefined}
        checked={true}
        className="components-checkbox-control__input"
        id="opt-1"
        onBlur={[Function]}
        onChange={[Function]}
        type="checkbox"
        value="1"
      />
      <label
        className="components-checkbox-control__label"
        htmlFor="inspector-checkbox-control-4"
      >
        One
      </label>
    </div>
  </div>
  <div
    className="components-base-control"
  >
    <div
      className="components-base-control__field"
    >
      <input
        aria-describedby="inspector-checkbox-control-5__help"
        checked={true}
        className="components-checkbox-control__input"
        id="opt-2"
        onBlur={[Function]}
        onChange={[Function]}
        type="checkbox"
        value="1"
      />
      <label
        className="components-checkbox-control__label"
        htmlFor="inspector-checkbox-control-5"
      >
        Two
      </label>
    </div>
    <p
      className="components-base-control__help"
      id="inspector-checkbox-control-5__help"
    >
      The Second Option
    </p>
  </div>
  <div
    className="components-base-control"
  >
    <div
      className="components-base-control__field"
    >
      <input
        aria-describedby={undefined}
        checked={true}
        className="components-checkbox-control__input"
        id="opt-3"
        onBlur={[Function]}
        onChange={[Function]}
        type="checkbox"
        value="1"
      />
      <label
        className="components-checkbox-control__label"
        htmlFor="inspector-checkbox-control-6"
      >
        Three
      </label>
    </div>
  </div>
</fieldset>
`;

exports[`fieldFactory Creates a number field 1`] = `
<div
  className="components-base-control caldera-field caldera-field-number"
>
  <div
    className="components-base-control__field"
  >
    <label
      className="components-base-control__label"
      htmlFor="inspector-text-control-3"
    >
      Number Field Label
    </label>
    <input
      aria-describedby="inspector-text-control-3__help"
      className="components-text-control__input"
      id="numberFieldId"
      max={10}
      min={1}
      onBlur={[Function]}
      onChange={[Function]}
      placeholder={undefined}
      type="number"
      value={3}
    />
  </div>
  <p
    className="components-base-control__help"
    id="inspector-text-control-3__help"
  >
    Number field description
  </p>
</div>
`;

exports[`fieldFactory Creates a radio field 1`] = `
<div
  className="components-base-control caldera-field caldera-field-radio components-radio-control"
>
  <div
    className="components-base-control__field"
  >
    <label
      className="components-base-control__label"
      htmlFor="inspector-radio-control-0"
    >
      Radio Field Label
    </label>
    <div
      className="components-radio-control__option"
    >
      <input
        aria-describedby="inspector-radio-control-0__help"
        checked={false}
        className="components-radio-control__input"
        id="inspector-radio-control-0-0"
        name="inspector-radio-control-0"
        onChange={[Function]}
        type="radio"
        value={0}
      />
      <label
        htmlFor="inspector-radio-control-0-0"
      >
        Zero
      </label>
    </div>
    <div
      className="components-radio-control__option"
    >
      <input
        aria-describedby="inspector-radio-control-0__help"
        checked={true}
        className="components-radio-control__input"
        id="inspector-radio-control-0-1"
        name="inspector-radio-control-0"
        onChange={[Function]}
        type="radio"
        value={1}
      />
      <label
        htmlFor="inspector-radio-control-0-1"
      >
        One
      </label>
    </div>
    <div
      className="components-radio-control__option"
    >
      <input
        aria-describedby="inspector-radio-control-0__help"
        checked={false}
        className="components-radio-control__input"
        id="inspector-radio-control-0-2"
        name="inspector-radio-control-0"
        onChange={[Function]}
        type="radio"
        value={2}
      />
      <label
        htmlFor="inspector-radio-control-0-2"
      >
        Two
      </label>
    </div>
  </div>
  <p
    className="components-base-control__help"
    id="inspector-radio-control-0__help"
  >
    Radio field description
  </p>
</div>
`;

exports[`fieldFactory Creates a text field 1`] = `
<div
  className="components-base-control caldera-field caldera-field-text"
>
  <div
    className="components-base-control__field"
  >
    <label
      className="components-base-control__label"
      htmlFor="inspector-text-control-0"
    >
      First Name
    </label>
    <input
      aria-describedby="inspector-text-control-0__help"
      className="components-text-control__input"
      id="firstName"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder={undefined}
      type="text"
      value="roy"
    />
  </div>
  <p
    className="components-base-control__help"
    id="inspector-text-control-0__help"
  >
    your first name
  </p>
</div>
`;

exports[`fieldFactory Creates an autocomplete field 1`] = `
<div
  className="components-base-control"
>
  <div
    className="components-base-control__field"
  >
    <label
      className="components-base-control__label"
      htmlFor="autoComplete1"
    >
      Options
    </label>
    <div
      className="components-form-token-field"
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyPress={[Function]}
      tabIndex="-1"
    >
      <label
        className="components-form-token-field__label"
        htmlFor="components-form-token-input-0"
      >
        Add item
      </label>
      <div
        className="caldera-field caldera-field-autocomplete components-form-token-field__input-container"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        tabIndex="-1"
      >
        <span
          className="components-form-token-field__token"
          onMouseEnter={undefined}
          onMouseLeave={undefined}
          title={undefined}
        >
          <span
            className="components-form-token-field__token-text"
            id="components-form-token-field__token-text-0"
          >
            <span
              className="screen-reader-text"
            >
              One (1 of 2)
            </span>
            <span
              aria-hidden="true"
            >
              One
            </span>
          </span>
          <button
            aria-describedby="components-form-token-field__token-text-0"
            aria-label="Remove item"
            className="components-button components-icon-button components-form-token-field__remove-token"
            disabled={undefined}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            type="button"
          >
            <svg
              aria-hidden="true"
              className="dashicon dashicons-dismiss"
              focusable="false"
              height={20}
              role="img"
              viewBox="0 0 20 20"
              width={20}
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M10 2c4.42 0 8 3.58 8 8s-3.58 8-8 8-8-3.58-8-8 3.58-8 8-8zm5 11l-3-3 3-3-2-2-3 3-3-3-2 2 3 3-3 3 2 2 3-3 3 3z"
              />
            </svg>
          </button>
        </span>
        <span
          className="components-form-token-field__token"
          onMouseEnter={undefined}
          onMouseLeave={undefined}
          title={undefined}
        >
          <span
            className="components-form-token-field__token-text"
            id="components-form-token-field__token-text-1"
          >
            <span
              className="screen-reader-text"
            >
              Three (2 of 2)
            </span>
            <span
              aria-hidden="true"
            >
              Three
            </span>
          </span>
          <button
            aria-describedby="components-form-token-field__token-text-1"
            aria-label="Remove item"
            className="components-button components-icon-button components-form-token-field__remove-token"
            disabled={undefined}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            type="button"
          >
            <svg
              aria-hidden="true"
              className="dashicon dashicons-dismiss"
              focusable="false"
              height={20}
              role="img"
              viewBox="0 0 20 20"
              width={20}
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M10 2c4.42 0 8 3.58 8 8s-3.58 8-8 8-8-3.58-8-8 3.58-8 8-8zm5 11l-3-3 3-3-2-2-3 3-3-3-2 2 3 3-3 3 2 2 3-3 3 3z"
              />
            </svg>
          </button>
        </span>
        <input
          aria-activedescendant={undefined}
          aria-autocomplete="list"
          aria-describedby="components-form-token-suggestions-howto-0"
          aria-expanded={false}
          aria-owns={undefined}
          autoCapitalize={undefined}
          autoComplete={undefined}
          className="components-form-token-field__input"
          disabled={false}
          id="components-form-token-input-0"
          onBlur={[Function]}
          onChange={[Function]}
          role="combobox"
          size={1}
          type="text"
          value=""
        />
      </div>
      <div
        className="screen-reader-text"
        id="components-form-token-suggestions-howto-0"
      >
        Separate with commas
      </div>
    </div>
  </div>
  <p
    className="components-base-control__help"
    id="autoComplete1__help"
  >
    Help text
  </p>
</div>
`;

exports[`fieldFactory Creates an checkbox field field 1`] = `
<div
  className="components-base-control"
>
  <div
    className="components-base-control__field"
  >
    <input
      aria-describedby="inspector-checkbox-control-0__help"
      checked={true}
      className="components-checkbox-control__input"
      id="checkboxFieldLabel"
      onBlur={[Function]}
      onChange={[Function]}
      type="checkbox"
      value="1"
    />
    <label
      className="components-checkbox-control__label"
      htmlFor="inspector-checkbox-control-0"
    >
      Checkbox Labe
    </label>
  </div>
  <p
    className="components-base-control__help"
    id="inspector-checkbox-control-0__help"
  >
    Checkbox field description
  </p>
</div>
`;

exports[`fieldFactory Creates an checkbox field set 1`] = `
<fieldset
  className="caldera-fieldset caldera-fieldset-checkboxes"
>
  <legend>
    Checkbox Field Set Label
  </legend>
  <div
    className="components-base-control"
  >
    <div
      className="components-base-control__field"
    >
      <input
        aria-describedby={undefined}
        checked={true}
        className="components-checkbox-control__input"
        id="opt-1"
        onBlur={[Function]}
        onChange={[Function]}
        type="checkbox"
        value="1"
      />
      <label
        className="components-checkbox-control__label"
        htmlFor="inspector-checkbox-control-1"
      >
        One
      </label>
    </div>
  </div>
  <div
    className="components-base-control"
  >
    <div
      className="components-base-control__field"
    >
      <input
        aria-describedby="inspector-checkbox-control-2__help"
        checked={true}
        className="components-checkbox-control__input"
        id="opt-2"
        onBlur={[Function]}
        onChange={[Function]}
        type="checkbox"
        value="1"
      />
      <label
        className="components-checkbox-control__label"
        htmlFor="inspector-checkbox-control-2"
      >
        Two
      </label>
    </div>
    <p
      className="components-base-control__help"
      id="inspector-checkbox-control-2__help"
    >
      The Second Option
    </p>
  </div>
  <div
    className="components-base-control"
  >
    <div
      className="components-base-control__field"
    >
      <input
        aria-describedby={undefined}
        checked={true}
        className="components-checkbox-control__input"
        id="opt-3"
        onBlur={[Function]}
        onChange={[Function]}
        type="checkbox"
        value="1"
      />
      <label
        className="components-checkbox-control__label"
        htmlFor="inspector-checkbox-control-3"
      >
        Three
      </label>
    </div>
  </div>
</fieldset>
`;

exports[`fieldFactory Creates an email field 1`] = `
<div
  className="components-base-control caldera-field caldera-field-email"
>
  <div
    className="components-base-control__field"
  >
    <label
      className="components-base-control__label"
      htmlFor="inspector-text-control-4"
    >
      Email Field Label
    </label>
    <input
      aria-describedby="inspector-text-control-4__help"
      className="components-text-control__input"
      id="emailFieldId"
      multiple={true}
      onBlur={[Function]}
      onChange={[Function]}
      placeholder={undefined}
      type="email"
      value="emai@email.com"
    />
  </div>
  <p
    className="components-base-control__help"
    id="inspector-text-control-4__help"
  >
    Email field description
  </p>
</div>
`;

exports[`fieldFactory Creates an fields-autocomplete field 1`] = `
<div
  className="components-base-control"
>
  <div
    className="components-base-control__field"
  >
    <label
      className="components-base-control__label"
      htmlFor="firstName"
    >
      First Name
    </label>
    <div
      className="components-form-token-field"
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyPress={[Function]}
      tabIndex="-1"
    >
      <label
        className="components-form-token-field__label"
        htmlFor="components-form-token-input-1"
      >
        Add item
      </label>
      <div
        className="caldera-field caldera-field-autocomplete components-form-token-field__input-container"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        tabIndex="-1"
      >
        <span
          className="components-form-token-field__token"
          onMouseEnter={undefined}
          onMouseLeave={undefined}
          title={undefined}
        >
          <span
            className="components-form-token-field__token-text"
            id="components-form-token-field__token-text-2"
          >
            <span
              className="screen-reader-text"
            >
              r (1 of 3)
            </span>
            <span
              aria-hidden="true"
            >
              r
            </span>
          </span>
          <button
            aria-describedby="components-form-token-field__token-text-2"
            aria-label="Remove item"
            className="components-button components-icon-button components-form-token-field__remove-token"
            disabled={undefined}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            type="button"
          >
            <svg
              aria-hidden="true"
              className="dashicon dashicons-dismiss"
              focusable="false"
              height={20}
              role="img"
              viewBox="0 0 20 20"
              width={20}
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M10 2c4.42 0 8 3.58 8 8s-3.58 8-8 8-8-3.58-8-8 3.58-8 8-8zm5 11l-3-3 3-3-2-2-3 3-3-3-2 2 3 3-3 3 2 2 3-3 3 3z"
              />
            </svg>
          </button>
        </span>
        <span
          className="components-form-token-field__token"
          onMouseEnter={undefined}
          onMouseLeave={undefined}
          title={undefined}
        >
          <span
            className="components-form-token-field__token-text"
            id="components-form-token-field__token-text-3"
          >
            <span
              className="screen-reader-text"
            >
              o (2 of 3)
            </span>
            <span
              aria-hidden="true"
            >
              o
            </span>
          </span>
          <button
            aria-describedby="components-form-token-field__token-text-3"
            aria-label="Remove item"
            className="components-button components-icon-button components-form-token-field__remove-token"
            disabled={undefined}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            type="button"
          >
            <svg
              aria-hidden="true"
              className="dashicon dashicons-dismiss"
              focusable="false"
              height={20}
              role="img"
              viewBox="0 0 20 20"
              width={20}
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M10 2c4.42 0 8 3.58 8 8s-3.58 8-8 8-8-3.58-8-8 3.58-8 8-8zm5 11l-3-3 3-3-2-2-3 3-3-3-2 2 3 3-3 3 2 2 3-3 3 3z"
              />
            </svg>
          </button>
        </span>
        <span
          className="components-form-token-field__token"
          onMouseEnter={undefined}
          onMouseLeave={undefined}
          title={undefined}
        >
          <span
            className="components-form-token-field__token-text"
            id="components-form-token-field__token-text-4"
          >
            <span
              className="screen-reader-text"
            >
              y (3 of 3)
            </span>
            <span
              aria-hidden="true"
            >
              y
            </span>
          </span>
          <button
            aria-describedby="components-form-token-field__token-text-4"
            aria-label="Remove item"
            className="components-button components-icon-button components-form-token-field__remove-token"
            disabled={undefined}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            type="button"
          >
            <svg
              aria-hidden="true"
              className="dashicon dashicons-dismiss"
              focusable="false"
              height={20}
              role="img"
              viewBox="0 0 20 20"
              width={20}
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M10 2c4.42 0 8 3.58 8 8s-3.58 8-8 8-8-3.58-8-8 3.58-8 8-8zm5 11l-3-3 3-3-2-2-3 3-3-3-2 2 3 3-3 3 2 2 3-3 3 3z"
              />
            </svg>
          </button>
        </span>
        <input
          aria-activedescendant={undefined}
          aria-autocomplete="list"
          aria-describedby="components-form-token-suggestions-howto-1"
          aria-expanded={false}
          aria-owns={undefined}
          autoCapitalize={undefined}
          autoComplete={undefined}
          className="components-form-token-field__input"
          disabled={false}
          id="components-form-token-input-1"
          onBlur={[Function]}
          onChange={[Function]}
          role="combobox"
          size={1}
          type="text"
          value=""
        />
      </div>
      <div
        className="screen-reader-text"
        id="components-form-token-suggestions-howto-1"
      >
        Separate with commas
      </div>
    </div>
  </div>
  <p
    className="components-base-control__help"
    id="firstName__help"
  >
    your first name
  </p>
</div>
`;

exports[`fieldFactory Creates an select field 1`] = `
<div
  className="components-base-control caldera-field caldera-field-select"
>
  <div
    className="components-base-control__field"
  >
    <label
      className="components-base-control__label"
      htmlFor="inspector-select-control-0"
    >
      Select Field Label
    </label>
    <select
      aria-describedby="inspector-select-control-0__help"
      className="components-select-control__input"
      id="selectFieldId"
      multiple={false}
      onChange={[Function]}
      value={1}
    >
      <option
        value={0}
      >
        Zero
      </option>
      <option
        value={1}
      >
        One
      </option>
      <option
        value={2}
      >
        Two
      </option>
    </select>
  </div>
  <p
    className="components-base-control__help"
    id="inspector-select-control-0__help"
  >
    Select field description
  </p>
</div>
`;

exports[`fieldFactory Creates an select field identified as dropdown 1`] = `
<div
  className="components-base-control caldera-field caldera-field-select"
>
  <div
    className="components-base-control__field"
  >
    <label
      className="components-base-control__label"
      htmlFor="inspector-select-control-1"
    >
      Select Field Label
    </label>
    <select
      aria-describedby="inspector-select-control-1__help"
      className="components-select-control__input"
      id="selectFieldId"
      multiple={false}
      onChange={[Function]}
      value={1}
    >
      <option
        value={0}
      >
        Zero
      </option>
      <option
        value={1}
      >
        One
      </option>
      <option
        value={2}
      >
        Two
      </option>
    </select>
  </div>
  <p
    className="components-base-control__help"
    id="inspector-select-control-1__help"
  >
    Select field description
  </p>
</div>
`;

exports[`fieldFactory should create textarea 1`] = `
<div
  className="components-base-control"
>
  <div
    className="components-base-control__field"
  >
    <label
      className="components-base-control__label"
      htmlFor="inspector-textarea-control-0"
    >
      Your message
    </label>
    <textarea
      aria-describedby="inspector-textarea-control-0__help"
      className="components-textarea-control__input"
      cols={8}
      id="textArea1"
      maxLength={500}
      minlength={10}
      onBlur={[Function]}
      onChange={[Function]}
      required={true}
      rows={5}
      value="The content of text area"
    />
  </div>
  <p
    className="components-base-control__help"
    id="inspector-textarea-control-0__help"
  >
    your first name
  </p>
</div>
`;

exports[`fieldFactory should create toggle 1`] = `
<div
  className="components-base-control"
>
  <div
    className="components-base-control__field"
  >
    <label
      className="components-base-control__label"
      htmlFor="toggleFieldLabel"
    >
      Toggle Label
    </label>
    <div
      className="components-base-control components-toggle-control"
    >
      <div
        className="components-base-control__field"
      >
        <span
          className="components-form-toggle is-checked"
        >
          <input
            aria-describedby={undefined}
            checked={true}
            className="components-form-toggle__input"
            id="inspector-toggle-control-0"
            onChange={[Function]}
            type="checkbox"
          />
          <span
            className="components-form-toggle__track"
          />
          <span
            className="components-form-toggle__thumb"
          />
          <svg
            aria-hidden="true"
            className="components-form-toggle__on"
            focusable="false"
            height="6"
            role="img"
            viewBox="0 0 2 6"
            width="2"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h2v6H0z"
            />
          </svg>
        </span>
        <label
          className="components-toggle-control__label"
          htmlFor="inspector-toggle-control-0"
        />
      </div>
    </div>
  </div>
  <p
    className="components-base-control__help"
    id="toggleFieldLabel__help"
  >
    Toggle field description
  </p>
</div>
`;
